Move seeking logic into a custom editor clock

smoogipoo opened the pull request:
* `Editor` now has an `EditorClock`.
* Moves the seek handling from `HitObjectComposer` to `EditorClock`.
* Cleans up a lot of testcases by encapsulating the clock construction/handling logic.
* Exposed as an `IAdjustableClock` to editor components through DI - if we need components to snap later on then we may expose it as the `EditorClock` itself, but as we discussed earlier this is not needed for now.

smoogipoo committed:
Give editor a custom clock to handle seeking

smoogipoo committed:
Fix + cleanup testcases

smoogipoo committed:
HitObjectComposer no longer needs a beat divisor

smoogipoo committed:
Fix Compose not binding to the editor's beat divisor

smoogipoo committed:
DI beat divisors to test cases with editor clocks

smoogipoo committed:
Remove re-instantiation of clock in EditorClockTestCase

smoogipoo committed:
Give the test case clock an accurate IsRunning value

peppy commented on osu.Game/Tests/Visual/EditorClockTestCase.cs:
```diff
@@ -0,0 +1,86 @@
+        {
+            base.Update();
+
+            // We don't have any explicit way to start/stop the track, but want a relatively accurate IsRunning state
```
i'm not sure this comment explains what's going on here so well... i can't understand it.

peppy commented on osu.Game/Screens/Edit/Components/Timelines/Summary/SummaryTimeline.cs:
```diff
@@ -17,8 +17,15 @@ namespace osu.Game.Screens.Edit.Components.Timelines.Summary
     {
+        private readonly IAdjustableClock adjustableClock;
+
+        public SummaryTimeline(IAdjustableClock adjustableClock)
```
what's the reasoning behind moving this out of DI?

smoogipoo committed:
Merge master into editor-clock

smoogipoo committed:
Move back to DI-ing adjustable clock into SummaryTimeline

smoogipoo committed:
Remove start/stop clock logic

Beatmap track shouldn't be start/stopped anyway - the IAdjustableClock should be DI'd in to perform the functionality.

peppy committed:
Merge branch 'master' into editor-clock

peppy commented on osu.Game/Screens/Edit/EditorClock.cs:
```diff
@@ -0,0 +1,109 @@
+            ControlPointInfo = controlPointInfo;
+        }
+
+        public bool SeekSnapped(double position)
```
probably best to xmldoc this (and hopefully this class too)

peppy committed:
Add some missing xmldoc

smoogipoo commented on osu.Game/Screens/Edit/EditorClock.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        /// <summary>
+        /// Seek to the closest valid snap value.
```
This should be something more like "Seek to the closest beat at a time."

smoogipoo commented on osu.Game/Screens/Edit/EditorClock.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        /// <summary>
+        /// Seeks the current time one beat-snapped beat-length backwards.
```
Not necessarily snapped (see param).

smoogipoo commented on osu.Game/Screens/Edit/EditorClock.cs:
```diff
@@ -0,0 +1,117 @@
+        public void SeekBackward(bool snapped = false) => seek(-1, snapped);
+
+        /// <summary>
+        /// Seeks the current time one beat-snapped beat-length forwards.
```
Not necessarily snapped (see param).

peppy committed:
Fix smoogipoo's comments
